src/librustc_typeck/check/coercion.rs

```diff
@@ -897,25 +896,61 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 /// let final_ty = coerce.complete(fcx);
 /// ```
 #[derive(Clone)] // (*)
-pub struct CoerceMany<'gcx: 'tcx, 'tcx> {
+pub struct CoerceMany<'gcx, 'tcx, 'exprs, E>
+    where 'gcx: 'tcx, E: 'exprs + AsCoercionSite,
+{
     expected_ty: Ty<'tcx>,
     final_ty: Option<Ty<'tcx>>,
-    expressions: Vec<&'gcx hir::Expr>,
+    expressions: Expressions<'gcx, 'exprs, E>,
+    pushed: usize,
+}
+
+/// The type of a `CoerceMany` that is storing up the expressions into
+/// a buffer. We use this in `check/mod.rs` for things like `break`.
+pub type DynamicCoerceMany<'gcx, 'tcx> = CoerceMany<'gcx, 'tcx, 'static, hir::Expr>;
+
+#[derive(Clone)] // (*)
+enum Expressions<'gcx, 'exprs, E>
+    where E: 'exprs + AsCoercionSite,
+{
+    Dynamic(Vec<&'gcx hir::Expr>),
+    UpFront(&'exprs [E]),
 }
 
 // (*) this is clone because `FnCtxt` is clone, but it seems dubious -- nmatsakis
```

Comment:
Can we remove `Clone` from `FnCtxt`? It's probably a mistake, we never clone contexts.